// socialreport-webservice/model/posts.js
'use strict';

var async = require('async');
var chalk = require('chalk');
var Promise = require('bluebird');
var socialClientHandler = require('../config/socialclient');
var socialclient = socialClientHandler();
var ObjectID = require('mongodb').ObjectID;

// User model
module.exports = function (app, callback) {

    app.db.collection('posts', function (err, collection) {

        var model = {

            getPostTimeline: function (accountid, channel, name, callback) {
                // change 'BSI_UK' to client

                switch (channel) {
                    case "twitter":
                        console.log("Getting Twitter posts for: " + name);

                        var options = {screen_name: name, user_id: name, count: 200};
                        var action = "statuses/user_timeline";

                        //var result = socialclient.getFollowers(action,accountName);
                        socialclient.getPostTimeline(channel, action, options).then(function (posts) {
                            console.log("resolved postTimeline");
                            console.log(posts.length);
                            //console.log(posts);
                            var account = {
                                account: {
                                    id: new ObjectID(accountid),
                                    preparedPosts: []
                                }
                            };
                            //var preparedPosts = [];

                            posts.forEach(function (post) {
                                account.account.preparedPosts.push({
                                    //account: new ObjectID(accountid),
                                    text: post.text,
                                    date: post.created_at
                                });
                            });

                            var results = Promise.all(account.account.preparedPosts);

                            results.then(function (data) {
/*
                                collection.insert(account, function (err, result) {
                                    if (err) {
                                        return callback(err);
                                    }
                                    callback(null, result);
                                });
*/                              console.log("acount to update in db: " + account.account.id);

                                collection.update(
                                    {"account.id": account.account.id},
                                    {account: account.account},
                                    {upsert: true},
                                    function (err, result) {
                                        callback(null, result);
                                    });

                                /*
                                 collection.findOneAndUpdate(
                                 {"account.id": account.account.id},
                                 {$set: {account: account.account}},
                                 {upsert: true, returnNewDocument : true}
                                 )
                                 */
                                //callback(null, data);
                            });
                            //callback(null, posts);
                        });


                        break;
                    case "facebook":
                        console.log("Getting Facebook posts for: " + name);
                        callback(null, [{text: "No Posts"}]);
                        break;
                    case "linkedin":
                        console.log("Getting LinkedIn posts for: " + name);
                        callback(null, [{text: "No Posts"}]);
                        break;
                    default:

                }

                /*
                 var posts = [{
                 text: "Discover how having access to over 90,000 internationally recognized #standards can help your organization. >>>> http://bit.ly/2gUYJet"
                 }, {
                 text: "How does Organisational Resilience link with Lean Six Sigma? > http://bit.ly/2fQzgRC"
                 }, {
                 text: "Find out how #standards can help your organization achieve excellence. >>>> http://bit.ly/2g2gAMx"
                 }];

                 callback(null, posts);
                 */
            },

            getPosts: function (account, callback) {
                console.log("In model posts getPosts");
                console.log("account: " + account);

                var id = new ObjectID(account);

                collection
                    .find({"account.id": id})
                    // .sort({followers_count: -1})
                    .toArray(function (err, results) {
                        if (err) {
                            return callback(err);
                        }
                        console.log("Results: " + JSON.stringify(results));
                        callback(null, results);
                    });
            }

        };
        callback(null, model);
    });
}
